### PR TITLE
git: Avoid resetting Git index when trees are identical

### PR DESCRIPTION
**Pull request by mlcui-corp**
This accounts for commit message-only changes which do not change the tree ID, as well as checkouts between commits with identical tree contents.

See #3748 for previous work on avoiding resetting HEAD when the *commits* are identical.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/git.rs [hunk 1]
```diff
@@ -933,9 +933,23 @@ pub fn reset_head(
         let git_head = mut_repo.view().git_head();
         let new_git_commit_id = Oid::from_bytes(first_parent_id.as_bytes()).unwrap();
         let new_git_commit = git_repo.find_commit(new_git_commit_id)?;
-        if git_head == &first_parent {
-            // `HEAD@git` already points to the correct commit, so we only need to reset
-            // the Git index. Only do so if it is non-empty (i.e. a user used `git add`).
+        let is_same_tree = if git_head == &first_parent {
+            true
+        } else if let Some(git_head_id) = git_head.as_normal() {
```
[the remark this answers is not in the document]
Comment (yuja) [reply lib/src/git.rs:938]: Maybe you can move `mut_repo.set_git_head_target(first_parent);` to the bottom?
It's common between root_id and non-root_id cases.

I think it's also fine to clone `git_head`.

File: lib/src/git.rs
Comment (yuja) [inline lib/src/git.rs:1]: (In commit message)
> git: Avoid resetting HEAD when trees are identical

Better to say "resetting Git index" or "staging area"? "Avoid resetting HEAD" sounds like bypassing `set_head_detached()`.

**Review by yuja — APPROVED** [review]

File: lib/src/git.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mlcui-corp) [reply lib/src/git.rs:943]: Done in #3774, which this PR now builds on top of.

**Comment by mlcui-corp** [comment]
DO NOT SUBMIT until #3774 is merged.

**Comment by mlcui-corp** [comment]
Please take another look, thanks! This should be much cleaner now.

**Review by yuja — APPROVED** [review]

**Pull request merged**